doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,286 @@
+accurately reflect the final implementation.
+
+[validation spike]: https://github.com/alphagov/search-api/pull/1768
+
+### Language
+
+We would like to keep the code for this project in Ruby as much as
+possible. Search API is written in Ruby and it is the primary language
+used at GOV.UK. Necessarily some machine learning code will need to be
+written in Python.
+
+### Machine learning library
+
+We have decided to [TensorFlow's ranking library][] for our
+implementation of Learning To Rank.
+
+TensorFlow is well supported, and the ranking library is under active development.
+
+We considered using the [elasticsearch-learning-to-rank][] plugin.
+However, AWS Elasticsearch does not permit us to install custom plugins.
+
+We also considered PyTorch. However, this would require writing much more
+custom machine learning code, as PyTorch don't currently have a ranking library.
+
+[TensorFlow's ranking library]: https://github.com/tensorflow/ranking
+[elasticsearch-learning-to-rank]: https://github.com/o19s/elasticsearch-learning-to-rank
+
+### Dockerised model
+
+We will use TensorFlow Serving to make a re-ranker model available
+to Search API on search machines.
+
+We will run a docker container (containing our pre-trained model) on
+search machines to which Search API can send requests.
+
+Should the docker container fail, Search API should be resilient to this,
+and fall back to the default ranking algorithm, with results not
+re-ranked by the model.
+
+### Training
+
+**Training the model** should happen on search machines on a schedule.
+An instance of Search API will train a model using the latest relevancy
+judgements stored in S3, and will then upload a new version of the model to S3.
+On a schedule all search machines will pull the latest version of the model
+and push the new version to the locally running TF Serving container.
+
+**Training data** is made up of relevancy judgements and document features.
+The judgements may be sourced from user behaviour (implicit signals
+of relevancy) perhaps with a Click Model; so far we have used explicit
+manual judgements.
+
+Acquiring relevancy judgements is a work in progress. Our next project
+will be to create a process by which we can acquire relevancy judgements
+on a rolling basis automatically.
+
+### Re-ranking requests
+
+**Using the model** to re-rank results will involve a user searching,
+Search API will perform the query for the top-k results (say 100),
+next Search API will make a request for new scores from the re-ranker
+for the results, finally Search API will provide the re-ranked top-n
+results to the user.
+
+We may see greater search query latency when using the model. Search
+API will need to pull more data out of elasticsearch and do more
+computation, in addition to making a request to the re-ranker.
+This is something we will monitor, as we don't want to provide a much
+slower search experience at a cost of improving relevance.
```

Comment:
Yes! I'm wondering about trying to introduce performance budgets.  We could mitigate some backend slowdown by improving the experience elsewhere - for example by rendering less on the page, or doing some optimisation of assets.  It's important that we can keep a track of this though

Reply:
Yes definitely. It's nice that the performance bottlenecks seem to be within our control.

Reply:
Do you have any thoughts on the performance budgets we could place on Search API response time prior to implementing LTR?